File: typescript/tests/integration/browser-screenshot.integration.test.ts
/* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access */
/**
 * Browser screenshot integration tests for TypeScript SDK
 * 
 * These tests verify the browser screenshot functionality by:
 * - Creating browser sessions with AgentBay
 * - Initializing browsers with various configurations
 * - Connecting to browsers via Playwright
 * - Taking screenshots with different options
 * - Validating screenshot data and error handling
 */

import { AgentBay, Browser, BrowserOptionClass, log } from '../../src';
import { chromium } from 'playwright';

function getTestApiKey(): string {
  const apiKey = process.env.AGENTBAY_API_KEY;
  if (!apiKey) {
    log("Warning: Using default API key. Set AGENTBAY_API_KEY environment variable for testing.");
    return "akm-xxx";
  }
  return apiKey as string;
}

function maskSecret(secret: string, visible = 4): string {
  if (!secret) return "";
  if (secret.length <= visible) return "*".repeat(secret.length);
  return "*".repeat(secret.length - visible) + secret.slice(-visible);
}

describe('Browser Screenshot Integration Tests', () => {
  let agentBay: any;
  let session: any;

  beforeEach(async () => {
    const apiKey = getTestApiKey();
    log("api_key =", maskSecret(apiKey));
    agentBay = new AgentBay({ apiKey });

    // Create a session
    log("Creating a new session for browser screenshot testing...");
    const sessionParam = {
      imageId: "browser_latest"
    };
    
    const result = await agentBay.create(sessionParam);
    
    if (!result.success) {
      log("⚠️ Session creation failed - probably due to resource limitations");
      log("Result data:", result.errorMessage || result);
      session = null; // Mark as failed
      return;
    }
    
    session = result.session;
    log(`Session created with ID: ${session.sessionId}`);
  });

  afterEach(async () => {
    if (session) {
      log("Cleaning up: Deleting the session...");
      try {
        await agentBay.delete(session);
      } catch (error: any) {
        log(`Warning: Error deleting session: ${error?.message || error}`);
      }
    }
  });

  test('Screenshot with valid page', async () => {
    if (!session) {
      log("⏭️ Skipping test - session creation failed");
      expect(true).toBe(true);
      return;
    }

    const browser = session.browser;
    expect(browser).toBeDefined();

    const initialized = await browser.initializeAsync({});
    expect(initialized).toBe(true);

    try {
      const endpointUrl = await browser.getEndpointUrl();
      log("endpoint_url:", endpointUrl);
      expect(endpointUrl).toBeDefined();

      // Connect with Playwright
      const playwrightBrowser = await chromium.connectOverCDP(endpointUrl);
      const context = playwrightBrowser.contexts()[0];
      const page = await context.newPage();

      // Navigate to a page
      await page.goto("https://www.aliyun.com", { timeout: 30000 });
      await page.waitForLoadState("domcontentloaded");

      // Take screenshot with default settings (full_page=false)
      try {
        const screenshotData = await browser.screenshot(page);
        // This should fail because we haven't implemented Playwright integration yet
        // but we're testing that the method properly indicates what needs to be done
        expect(true).toBe(false); // Should not reach here
      } catch (error: any) {
        expect(error.message).toContain("Screenshot functionality requires Playwright TypeScript integration");
      }

      await page.close();
      await playwrightBrowser.close();
    } catch (error: any) {
      log("Expected: Browser operations might fail in test environment:", error?.message || error);
    }
  }, 60000);

  test('Screenshot with full page option', async () => {
    if (!session) {
      log("⏭️ Skipping test - session creation failed");
      expect(true).toBe(true);
      return;
    }

    const browser = session.browser;
    expect(browser).toBeDefined();

    const initialized = await browser.initializeAsync({});
    expect(initialized).toBe(true);

    try {
      const endpointUrl = await browser.getEndpointUrl();
      log("endpoint_url:", endpointUrl);
      expect(endpointUrl).toBeDefined();

      // Connect with Playwright
      const playwrightBrowser = await chromium.connectOverCDP(endpointUrl);
      const context = playwrightBrowser.contexts()[0];
      const page = await context.newPage();

      // Navigate to a page
      await page.goto("https://www.aliyun.com", { timeout: 30000 });
      await page.waitForLoadState("domcontentloaded");

      // Take screenshot with full_page=true
      try {
        const screenshotData = await browser.screenshot(page, true);
        // This should fail because we haven't implemented Playwright integration yet
        // but we're testing that the method properly indicates what needs to be done
        expect(true).toBe(false); // Should not reach here
      } catch (error: any) {
        expect(error.message).toContain("Screenshot functionality requires Playwright TypeScript integration");
      }

      await page.close();
      await playwrightBrowser.close();
    } catch (error: any) {
      log("Expected: Browser operations might fail in test environment:", error?.message || error);
    }
  }, 60000);

  test('Screenshot with custom options', async () => {
    if (!session) {
      log("⏭️ Skipping test - session creation failed");
      expect(true).toBe(true);
      return;
    }

    const browser = session.browser;
    expect(browser).toBeDefined();

    const initialized = await browser.initializeAsync({});
    expect(initialized).toBe(true);

    try {
      const endpointUrl = await browser.getEndpointUrl();
      log("endpoint_url:", endpointUrl);
      expect(endpointUrl).toBeDefined();

      // Connect with Playwright
      const playwrightBrowser = await chromium.connectOverCDP(endpointUrl);
      const context = playwrightBrowser.contexts()[0];
      const page = await context.newPage();

      // Navigate to a page
      await page.goto("https://www.aliyun.com", { timeout: 30000 });
      await page.waitForLoadState("domcontentloaded");

      // Take screenshot with custom options
      try {
        const screenshotData = await browser.screenshot(
          page,
          false, // full_page
          {
            type: "jpeg",
            quality: 80,
            timeout: 30000
          }
        );
        // This should fail because we haven't implemented Playwright integration yet
        // but we're testing that the method properly indicates what needs to be done
        expect(true).toBe(false); // Should not reach here
      } catch (error: any) {
        expect(error.message).toContain("Screenshot functionality requires Playwright TypeScript integration");
      }

      await page.close();
      await playwrightBrowser.close();
    } catch (error: any) {
      log("Expected: Browser operations might fail in test environment:", error?.message || error);
    }
  }, 60000);

  test('Screenshot without browser initialization', async () => {
    if (!session) {
      log("⏭️ Skipping test - session creation failed");
      expect(true).toBe(true);
      return;
    }

    // Create a new browser instance that is not initialized
    const uninitializedBrowser = session.browser;

    try {
      const endpointUrl = await session.browser.getEndpointUrl();
      log("endpoint_url:", endpointUrl);
      expect(endpointUrl).toBeDefined();

      // Connect with Playwright
      const playwrightBrowser = await chromium.connectOverCDP(endpointUrl);
      const context = playwrightBrowser.contexts()[0];
      const page = await context.newPage();

      // Navigate to a page
      await page.goto("https://www.aliyun.com", { timeout: 30000 });
      await page.waitForLoadState("domcontentloaded");

      // Attempt to take screenshot with uninitialized browser
      // Manually set browser as uninitialized for testing
      (uninitializedBrowser as any)._initialized = false;
      
      try {
        await uninitializedBrowser.screenshot(page);
        expect(true).toBe(false); // Should not reach here
      } catch (error: any) {
        expect(error.message).toContain("Browser must be initialized before calling screenshot");
      }

      await page.close();
      await playwrightBrowser.close();
    } catch (error: any) {
      log("Expected: Browser operations might fail in test environment:", error?.message || error);
    }
  }, 60000);

  test('Screenshot with multiple pages', async () => {
    if (!session) {
      log("⏭️ Skipping test - session creation failed");
      expect(true).toBe(true);
      return;
    }

    const browser = session.browser;
    expect(browser).toBeDefined();

    const initialized = await browser.initializeAsync({});
    expect(initialized).toBe(true);

    const urls = [
      "https://www.aliyun.com",
      "https://www.taobao.com"
    ];

    try {
      const endpointUrl = await browser.getEndpointUrl();
      log("endpoint_url:", endpointUrl);
      expect(endpointUrl).toBeDefined();

      // Connect with Playwright
      const playwrightBrowser = await chromium.connectOverCDP(endpointUrl);
      const context = playwrightBrowser.contexts()[0];

      for (let i = 0; i < urls.length; i++) {
        const page = await context.newPage();

        // Navigate to a page
        await page.goto(urls[i], { timeout: 30000 });
        await page.waitForLoadState("domcontentloaded");

        // Take screenshot with default settings
        try {
          const screenshotData = await browser.screenshot(page);
          // This should fail because we haven't implemented Playwright integration yet
          // but we're testing that the method properly indicates what needs to be done
          expect(true).toBe(false); // Should not reach here
        } catch (error: any) {
          expect(error.message).toContain("Screenshot functionality requires Playwright TypeScript integration");
        }

        await page.close();
        log(`✅ Screenshot ${i + 1} captured for ${urls[i]}`);
      }

      await playwrightBrowser.close();
    } catch (error: any) {
      log("Expected: Browser operations might fail in test environment:", error?.message || error);
    }
  }, 60000);

  test('Screenshot performance', async () => {
    if (!session) {
      log("⏭️ Skipping test - session creation failed");
      expect(true).toBe(true);
      return;
    }

    const browser = session.browser;
    expect(browser).toBeDefined();

    const initialized = await browser.initializeAsync({});
    expect(initialized).toBe(true);

    try {
      const endpointUrl = await browser.getEndpointUrl();
      log("endpoint_url:", endpointUrl);
      expect(endpointUrl).toBeDefined();

      // Connect with Playwright
      const playwrightBrowser = await chromium.connectOverCDP(endpointUrl);
      const context = playwrightBrowser.contexts()[0];
      const page = await context.newPage();

      // Navigate to a page
      await page.goto("https://www.aliyun.com", { timeout: 30000 });
      await page.waitForLoadState("domcontentloaded");

      // Measure screenshot time
      const startTime = Date.now();

      // Take screenshot
      try {
        const screenshotData = await browser.screenshot(page);
        // This should fail because we haven't implemented Playwright integration yet
        // but we're testing that the method properly indicates what needs to be done
        expect(true).toBe(false); // Should not reach here
      } catch (error: any) {
        expect(error.message).toContain("Screenshot functionality requires Playwright TypeScript integration");
      }

      const endTime = Date.now();
      const duration = (endTime - startTime) / 1000; // Convert to seconds

      await page.close();
      await playwrightBrowser.close();

      log(`✅ Screenshot attempted in ${duration.toFixed(2)} seconds`);

      // Performance check (should complete within reasonable time even if it fails)
      expect(duration).toBeLessThan(30);
    } catch (error: any) {
      log("Expected: Browser operations might fail in test environment:", error?.message || error);
    }
  }, 60000);
});

File: golang/tests/pkg/integration/browser_screenshot_integration_test.go
package integration

import (
	"os"
	"testing"
	"time"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay/browser"
	"github.com/playwright-community/playwright-go"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestBrowserScreenshot_Integration(t *testing.T) {
	// Skip if no API key
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Skip("Skipping integration test: AGENTBAY_API_KEY not set")
	}

	// Create AgentBay client
	agentBay, err := agentbay.NewAgentBay(apiKey)
	require.NoError(t, err)
	require.NotNil(t, agentBay)

	// Create session with browser image
	params := &agentbay.CreateSessionParams{
		ImageId: "browser_latest",
	}

	sessionResult, err := agentBay.Create(params)
	require.NoError(t, err)
	require.NotNil(t, sessionResult)
	require.True(t, sessionResult.Success)
	require.NotNil(t, sessionResult.Session)

	session := sessionResult.Session
	defer func() {
		// Clean up session
		_, err := agentBay.Delete(session, false)
		if err != nil {
			t.Logf("Warning: Failed to delete session: %v", err)
		}
	}()

	// Initialize browser
	option := browser.NewBrowserOption()
	success, err := session.Browser.Initialize(option)
	require.NoError(t, err)
	require.True(t, success)
	require.True(t, session.Browser.IsInitialized())

	// Get browser endpoint URL
	endpointURL, err := session.Browser.GetEndpointURL()
	require.NoError(t, err)
	require.NotEmpty(t, endpointURL)

	t.Run("Screenshot with valid page", func(t *testing.T) {
		// Start Playwright
		pw, err := playwright.Run()
		require.NoError(t, err)
		defer func() {
			_ = pw.Stop()
		}()

		// Connect to browser
		browserInstance, err := pw.Chromium.ConnectOverCDP(endpointURL)
		require.NoError(t, err)
		defer func() {
			_ = browserInstance.Close()
		}()

		context := browserInstance.Contexts()[0]
		page, err := context.NewPage()
		require.NoError(t, err)

		// Navigate to a page
		_, err = page.Goto("https://httpbin.org/html", playwright.PageGotoOptions{
			Timeout: playwright.Float(10000),
		})
		require.NoError(t, err)

		err = page.WaitForLoadState(playwright.PageWaitForLoadStateOptions{
			State: playwright.LoadStateDomcontentloaded,
		})
		require.NoError(t, err)

		// Take screenshot with default settings (full_page=false)
		screenshotData, err := session.Browser.Screenshot(page, nil)
		// Note: This should now work with Playwright integration
		assert.NoError(t, err)
		assert.NotNil(t, screenshotData)
		assert.Greater(t, len(screenshotData), 0)

		// Close page to free resources
		err = page.Close()
		assert.NoError(t, err)
	})

	t.Run("Screenshot with full page option", func(t *testing.T) {
		// Start Playwright
		pw, err := playwright.Run()
		require.NoError(t, err)
		defer func() {
			_ = pw.Stop()
		}()

		// Connect to browser
		browserInstance, err := pw.Chromium.ConnectOverCDP(endpointURL)
		require.NoError(t, err)
		defer func() {
			_ = browserInstance.Close()
		}()

		context := browserInstance.Contexts()[0]
		page, err := context.NewPage()
		require.NoError(t, err)

		// Navigate to a page
		_, err = page.Goto("https://httpbin.org/html", playwright.PageGotoOptions{
			Timeout: playwright.Float(10000),
		})
		require.NoError(t, err)

		err = page.WaitForLoadState(playwright.PageWaitForLoadStateOptions{
			State: playwright.LoadStateDomcontentloaded,
		})
		require.NoError(t, err)

		// Take screenshot with full_page=true
		options := &browser.ScreenshotOptions{
			FullPage: true,
		}
		screenshotData, err := session.Browser.Screenshot(page, options)
		// Note: This should now work with Playwright integration
		assert.NoError(t, err)
		assert.NotNil(t, screenshotData)
		assert.Greater(t, len(screenshotData), 0)

		// Close page to free resources
		err = page.Close()
		assert.NoError(t, err)
	})

	t.Run("Screenshot with custom options", func(t *testing.T) {
		// Start Playwright
		pw, err := playwright.Run()
		require.NoError(t, err)
		defer func() {
			_ = pw.Stop()
		}()

		// Connect to browser
		browserInstance, err := pw.Chromium.ConnectOverCDP(endpointURL)
		require.NoError(t, err)
		defer func() {
			_ = browserInstance.Close()
		}()

		context := browserInstance.Contexts()[0]
		page, err := context.NewPage()
		require.NoError(t, err)

		// Navigate to a page
		_, err = page.Goto("https://httpbin.org/html", playwright.PageGotoOptions{
			Timeout: playwright.Float(10000),
		})
		require.NoError(t, err)

		err = page.WaitForLoadState(playwright.PageWaitForLoadStateOptions{
			State: playwright.LoadStateDomcontentloaded,
		})
		require.NoError(t, err)

		// Take screenshot with custom options
		options := &browser.ScreenshotOptions{
			FullPage: false,
			Type:     "jpeg",
			Quality:  80,
			Timeout:  30000,
		}
		screenshotData, err := session.Browser.Screenshot(page, options)
		// Note: This should now work with Playwright integration
		assert.NoError(t, err)
		assert.NotNil(t, screenshotData)
		assert.Greater(t, len(screenshotData), 0)

		// Close page to free resources
		err = page.Close()
		assert.NoError(t, err)
	})

	t.Run("Screenshot without browser initialization", func(t *testing.T) {
		// Create a new browser instance that is not initialized
		uninitializedBrowser := browser.NewBrowser(session)

		// Start Playwright
		pw, err := playwright.Run()
		require.NoError(t, err)
		defer func() {
			_ = pw.Stop()
		}()

		// Connect to browser
		browserInstance, err := pw.Chromium.ConnectOverCDP(endpointURL)
		require.NoError(t, err)
		defer func() {
			_ = browserInstance.Close()
		}()

		context := browserInstance.Contexts()[0]
		page, err := context.NewPage()
		require.NoError(t, err)

		// Attempt to take screenshot with uninitialized browser
		screenshotData, err := uninitializedBrowser.Screenshot(page, nil)
		assert.Error(t, err)
		assert.Contains(t, err.Error(), "browser must be initialized before calling screenshot")
		assert.Nil(t, screenshotData)

		// Close page to free resources
		err = page.Close()
		assert.NoError(t, err)
	})

	t.Run("Screenshot with multiple pages", func(t *testing.T) {
		urls := []string{
			"https://httpbin.org/html",
			"https://httpbin.org/json",
		}

		// Start Playwright
		pw, err := playwright.Run()
		require.NoError(t, err)
		defer func() {
			_ = pw.Stop()
		}()

		// Connect to browser
		browserInstance, err := pw.Chromium.ConnectOverCDP(endpointURL)
		require.NoError(t, err)
		defer func() {
			_ = browserInstance.Close()
		}()

		context := browserInstance.Contexts()[0]

		for i, url := range urls {
			page, err := context.NewPage()
			require.NoError(t, err)

			// Navigate to a page
			_, err = page.Goto(url, playwright.PageGotoOptions{
				Timeout: playwright.Float(30000),
			})
			require.NoError(t, err)

			err = page.WaitForLoadState(playwright.PageWaitForLoadStateOptions{
				State: playwright.LoadStateDomcontentloaded,
			})
			require.NoError(t, err)

			// Take screenshot with default settings
			screenshotData, err := session.Browser.Screenshot(page, nil)
			// Note: This should now work with Playwright integration
			assert.NoError(t, err)
			assert.NotNil(t, screenshotData)
			assert.Greater(t, len(screenshotData), 0)

			// Close page
			err = page.Close()
			assert.NoError(t, err)

			t.Logf("✅ Screenshot %d captured for %s", i+1, url)
		}
	})

	t.Run("Screenshot performance", func(t *testing.T) {
		// Start Playwright
		pw, err := playwright.Run()
		require.NoError(t, err)
		defer func() {
			_ = pw.Stop()
		}()

		// Connect to browser
		browserInstance, err := pw.Chromium.ConnectOverCDP(endpointURL)
		require.NoError(t, err)
		defer func() {
			_ = browserInstance.Close()
		}()

		context := browserInstance.Contexts()[0]
		page, err := context.NewPage()
		require.NoError(t, err)

		// Navigate to a page
		_, err = page.Goto("https://httpbin.org/html", playwright.PageGotoOptions{
			Timeout: playwright.Float(10000),
		})
		require.NoError(t, err)

		err = page.WaitForLoadState(playwright.PageWaitForLoadStateOptions{
			State: playwright.LoadStateDomcontentloaded,
		})
		require.NoError(t, err)

		// Measure screenshot time
		startTime := time.Now()

		// Take screenshot
		screenshotData, err := session.Browser.Screenshot(page, nil)
		// Note: This should now work with Playwright integration
		assert.NoError(t, err)
		assert.NotNil(t, screenshotData)
		assert.Greater(t, len(screenshotData), 0)

		endTime := time.Now()
		duration := endTime.Sub(startTime)

		t.Logf("✅ Screenshot attempted in %v", duration)

		// Close page to free resources
		err = page.Close()
		assert.NoError(t, err)

		// Performance check (should complete within reasonable time even if it fails)
		assert.Less(t, duration, 30*time.Second, "Screenshot should complete within 30 seconds")
	})
}
